library/std/src/thread/mod.rs

```diff
@@ -718,7 +720,25 @@ pub(crate) fn set_current(thread: Thread) {
 /// In contrast to the public `current` function, this will not panic if called
 /// from inside a TLS destructor.
 pub(crate) fn try_current() -> Option<Thread> {
-    CURRENT.try_with(|current| current.get_or_init(|| Thread::new_unnamed()).clone()).ok()
+    CURRENT
+        .try_with(|current| {
+            let thread = current.get_or_init(Thread::new_unnamed).clone();
+            CURRENT_ID.set(Some(thread.id()));
+            thread
+        })
+        .ok()
+}
+
+/// Gets the id of the thread that invokes it.
+///
+/// If called from inside a TLS destructor and the thread was never
+/// assigned an id, returns `None`.
+#[inline]
+pub(crate) fn try_current_id() -> Option<ThreadId> {
+    if CURRENT_ID.get().is_none() {
+        let _ = try_current();
```

Comment:
This will access `CURRENT`, is that really what we want? I thought the goal is to not touch `CURRENT` at all in this function.

Reply:
It'll only access `CURRENT` at most once, when the `CURRENT_ID` is unset. In that case, accessing `CURRENT` is unavoidable, because one way or another you need to access the thread id. `try_current` then initializes the `CURRENT_ID` so that subsequent `try_current_id()` calls don't need to call `try_current()` again.

Reply:
Right, the entire lifecycle here is kind of subtle... and there are no comments to help the reader.

There's also the invariant that CURRENT and CURRENT_ID are always initialized together, I think?

Reply:
Yeah, they're always initialized together (but `CURRENT_ID` stays initialized even for TLS dtors, unlike `CURRENT`). I mostly just copied @CAD97's code from #123458, but I can definitely add some comments later.